cores/esp8266/WString.cpp

```diff
@@ -252,7 +260,22 @@ String & String::operator = (const __FlashStringHelper *pstr)
 // /*********************************************/
 
 unsigned char String::concat(const String &s) {
-    return concat(s.buffer, s.len);
+    // Special case if we're concatting ourself (s += s;) since we may end up
+    // realloc'ing the buffer and moving s.buffer in the method called
+    if (&s == this) {
+        unsigned int newlen = 2 * len;
+        if (!s.buffer)
+            return 0;
+        if (s.len == 0)
+            return 1;
+        if (!reserve(newlen))
+            return 0;
+        memcpy(s.buffer + len, s.buffer, len);
```

Comment:
If len includes the null term, then this will start copying after the first string's null term. If not, then this doesn't assure a null term result.

Reply:
Len doesn't include the null.  But this sequence does guarantee a 0-termination because reserve 0-fills the buffer on allocation...
https://github.com/earlephilhower/Arduino/blob/713faf8fe665d5e50f4b10618fdf584853d26052/cores/esp8266/WString.cpp#L161

Reply:
How very inefficient... Every time I look at this implementation I want to migrate to a std::string-based one.
In that case, approving.